src/NHapi.Base/Parser/PipeParser.cs

```diff
@@ -24,885 +24,960 @@
+                if (comps.Length >= 3)
+                {
+                    messageStructure = comps[2];
+                }
+                else if (comps.Length > 0 && comps[0] != null && comps[0].Equals("ACK"))
+                {
+                    messageStructure = "ACK";
+                }
+                else if (comps.Length == 2)
+                {
+                    explicityDefined = false;
+                    messageStructure = comps[0] + "_" + comps[1];
+                }
+
+                /*else if (comps.length == 1 && comps[0] != null && comps[0].equals("ACK")) {
+                 messageStructure = "ACK"; //it's common for people to only populate component 1 in an ACK msg
+                 }*/
+                else
+                {
+                    var buf = new StringBuilder("Can't determine message structure from MSH-9: ");
+                    buf.Append(wholeFieldNine);
+                    if (comps.Length < 3)
+                    {
+                        buf.Append(" HINT: there are only ");
+                        buf.Append(comps.Length);
+                        buf.Append(" of 3 components present");
+                    }
+
+                    throw new HL7Exception(buf.ToString(), ErrorCode.UNSUPPORTED_MESSAGE_TYPE);
+                }
+            }
+            catch (IndexOutOfRangeException e)
+            {
+                throw new HL7Exception("Can't find message structure (MSH-9-3): " + e.Message, ErrorCode.UNSUPPORTED_MESSAGE_TYPE);
+            }
+
+            return new MessageStructure(messageStructure, explicityDefined);
+        }
+
+        /// <summary>
+        /// A struct for holding a message class string and a boolean indicating whether it
+        /// was defined explicitly.
+        /// </summary>
+        private class MessageStructure
+        {
+            public MessageStructure(string theMessageStructure, bool isExplicitlyDefined)
+            {
+                Structure = theMessageStructure;
+                ExplicitlyDefined = isExplicitlyDefined;
+            }
+
+            public string Structure { get; }
+
+            public bool ExplicitlyDefined { get; }
+        }
+
+        private class AnonymousClassPredicate : FilterIterator.IPredicate
+        {
+            private PipeParser enclosingInstance;
+
+            public AnonymousClassPredicate(PipeParser enclosingInstance)
+            {
+                InitBlock(enclosingInstance);
+            }
+
+            public PipeParser Enclosing_Instance
+            {
+                get { return enclosingInstance; }
+            }
+
+            public virtual bool evaluate(object obj)
+            {
+                return Evaluate(obj);
+            }
+
+            public virtual bool Evaluate(object obj)
+            {
+                return typeof(ISegment).IsAssignableFrom(obj.GetType());
+            }
+
+            private void InitBlock(PipeParser enclosingInstance)

```

Comment:
not sure what the point of InitBlock is...
Can we remove it and move the assignment to the constructor?

Reply:
done